### PR TITLE
DeviceManager: Enable detection of Ping360 over ethernet

### LINKED ISSUE #601: ping360: implement ip address discovery via udp broadcast

### PR DESCRIPTION
**Pull request by Williangalvani**
Screenshot from 2019-07-31 17-16-02

fix #601

### CODE REVIEW

**Comment by Williangalvani** [comment]
There is a small issue. It keeps alternating between green and red:
ping360 intermitent

File: src/devicemanager/devicemanager.cpp [hunk 1]
```diff
@@ -59,12 +61,16 @@ void DeviceManager::append(const LinkConfiguration& linkConf, const QString& dev
 void DeviceManager::startDetecting()
 {
     qCDebug(DEVICEMANAGER) << "Start protocol detector service.";
+    _broadcastTimer.start(1000);
+    _broadcastSocket.bind(QHostAddress::Any, 1048);
```
Comment (patrickelectric) [inline src/devicemanager/devicemanager.cpp]: Thy is this in device manager and no in protocoldetector ?
DeviceManager is just a class that manages the device creation from the links that are possible to connect, everything related to the discovery process should be in protocol detector

Comment (patrickelectric) [inline src/devicemanager/devicemanager.cpp]: Do we need to use `bind` ? From what we talked about bind is not necessary if the other side is a server and it'll probably be.

File: src/devicemanager/devicemanager.cpp [hunk 2]
```diff
@@ -147,6 +153,28 @@ DeviceManager* DeviceManager::self()
     return self;
 }
 
+void DeviceManager::doBroadcast()
+{
+    QByteArray datagram = "Discovery";
```
Comment (patrickelectric) [inline src/devicemanager/devicemanager.cpp]: Why are you sending a "Discovery" string ?

File: src/devicemanager/devicemanager.cpp [hunk 3]
```diff
@@ -147,6 +153,28 @@ DeviceManager* DeviceManager::self()
...
+    _broadcastSocket.writeDatagram(datagram, QHostAddress::Broadcast, 30303);
```
Comment (patrickelectric) [inline src/devicemanager/devicemanager.cpp]: Why we are using port 30303 ?

File: src/devicemanager/devicemanager.h [hunk 4]
```diff
@@ -3,6 +3,8 @@
 #include <QAbstractListModel>
 #include <QLoggingCategory>
 #include <QThread>
+#include <QUdpSocket>
+#include <QTimer>
```
Comment (patrickelectric) [inline src/devicemanager/devicemanager.h]: Alphabetic order.

File: src/devicemanager/devicemanager.cpp (hunk 1 above)
Comment (Williangalvani) [reply src/devicemanager/devicemanager.cpp]: My reasoning was that we don't need to detect the protocol. We know it is a Ping360. I can move it over, but I vote for renaming ProtocolDetector to DeviceDetector or something like that =]

Comment (Williangalvani) [reply src/devicemanager/devicemanager.cpp]: We should need it, we are doing a broadcast, there is no connection involved. We say "Yo, any Ping360 out there?" then they all reply to the message.
That said, I just tested and it worked, so I'm taking it out ¯\_(ツ)_/¯

File: src/devicemanager/devicemanager.cpp (hunk 2 above)
Comment (Williangalvani) [reply src/devicemanager/devicemanager.cpp]: This is as documented in the "Communications Protocol v1.7.pdf"

File: src/devicemanager/devicemanager.cpp (hunk 3 above)
Comment (Williangalvani) [reply src/devicemanager/devicemanager.cpp]: This is as documented in the "Communications Protocol v1.7.pdf"

File: src/devicemanager/devicemanager.cpp [hunk 5]
```diff
@@ -147,6 +152,42 @@ DeviceManager* DeviceManager::self()
...
+}
+
+void DeviceManager::processBroadcastResponses()
+{
+    /*
+        We broadcast a "Discovery" message to 255.255.255.255 and port 30303 as
+        described in the communications manual.
+        The devices replies with IP and MAC addres in this format:
```
Comment (patrickelectric) [inline src/devicemanager/devicemanager.cpp]: Can something like this comment be add in the function description ?

**Comment by Williangalvani** [comment]
You are right, it is not clear that it is for Ping360 only. I'll get it into it's own class tomorrow. Maybe Ping360Udp<Detector|Scanner|Finder> (hopefully I come up with a better name during my sleep).

File: src/devicemanager/devicemanager.cpp (hunk 1 above)
Comment (patrickelectric) [reply src/devicemanager/devicemanager.cpp]: Indeed, We should rework and make it more abstract and clear what is the paper of each class.
Also, we should think more about the protocol detection side of things, allowing multiple protocols, different proprietary sensors and etc.

**Comment by Williangalvani** [comment]
@patrickelectric I moved it to a class

File: src/devicemanager/devicemanager.h [hunk 6]
```diff
@@ -175,6 +176,7 @@ class DeviceManager : public QAbstractListModel
 
     QSharedPointer<Sensor> _primarySensor;
     ProtocolDetector* _detector;
+    Ping360EthernetFinder* _ping360EthernetFinder;
```
Comment (patrickelectric) [inline src/devicemanager/devicemanager.h]: Is there a need for this to be a pointer ?

File: src/sensor/ping360ethernetfinder.cpp [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    while(_broadcastSocket.hasPendingDatagrams()) {
+        QHostAddress sender;
+        QByteArray datagram;
+        datagram.resize(_broadcastSocket.pendingDatagramSize());
+        _broadcastSocket.readDatagram(datagram.data(), datagram.size(), &sender);
+        // Make sure we have an IPV4 address, and not something like "::ffff:192.168.1.1"
+        sender = QHostAddress(sender.toIPv4Address());
+        // Basic validation to check we didn't get some lost packet from wherever
+        if(datagram.contains("PING360")) {
+            emit availableLinkFound({LinkType::Udp, {sender.toString(), "12345"}, "BlueRov2 Ping360 Port", PingDeviceType::PING360},
```
Comment (patrickelectric) [inline src/sensor/ping360ethernetfinder.cpp]: This is not BlueRov2 Ping360 port, the description should be fixed.

File: src/devicemanager/devicemanager.cpp [hunk 8]
```diff
@@ -21,6 +22,9 @@ DeviceManager::DeviceManager() :
         qCDebug(DEVICEMANAGER) << "Available devices:" << links;
         updateAvailableConnections(links);
     });
+    connect(_ping360EthernetFinder, &Ping360EthernetFinder::availableLinkFound, this, [this](auto link, auto name) {
+        append(link, name);
```
Comment (patrickelectric) [inline src/devicemanager/devicemanager.cpp]: do you need this lambda ? It appears that you can connect with append directly.

File: src/devicemanager/devicemanager.h (hunk 6 above)
Comment (Williangalvani) [reply src/devicemanager/devicemanager.h]: nope, fixed.

File: src/devicemanager/devicemanager.h [hunk 9]
```diff
@@ -6,6 +6,7 @@
 
 #include "abstractlinknamespace.h"
 #include "protocoldetector.h"
+#include "ping360ethernetfinder.h"
```
Comment (patrickelectric) [inline src/devicemanager/devicemanager.h]: alphabetic order

Comment (Williangalvani) [reply src/devicemanager/devicemanager.h]: done

**Pull request merged**